### PR TITLE
SONARJS-858 Resolve properties of built-in constructors

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BooleanBuiltInProperties.java [hunk 1]
```diff
@@ -22,38 +22,22 @@
...
+      .put("valueOf", method(Constraint.BOOLEAN_PRIMITIVE, BuiltInProperty.EMPTY))
 
       .build();
-  }
 
-  @Override
-  Map<String, Constraint> getPropertiesConstraints() {
-    return ImmutableMap.of();
-  }
+  public static final Map<String, BuiltInProperty> PROPERTIES = ImmutableMap.of();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BooleanBuiltInProperties.java]: MINOR Make this member "protected". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java [hunk 2]
```diff
@@ -40,16 +41,27 @@
...
+  public static final BuiltInObjectSymbolicValue DATE_PROTOTYPE = new BuiltInObjectSymbolicValue(DateBuiltInProperties.PROTOTYPE_PROPERTIES, OBJECT_PROTOTYPE, Constraint.DATE);
+  public static final BuiltInObjectSymbolicValue REGEXP_PROTOTYPE = new BuiltInObjectSymbolicValue(RegexpBuiltInProperties.PROTOTYPE_PROPERTIES, OBJECT_PROTOTYPE, Constraint.REGEXP);
+
+  public static final BuiltInObjectSymbolicValue MATH = new BuiltInObjectSymbolicValue(MathBuiltInProperties.PROPERTIES, OBJECT_PROTOTYPE, Constraint.OTHER_OBJECT);
+
+  public static final BuiltInConstructorSymbolicValue OBJECT = new BuiltInConstructorSymbolicValue(ObjectBuiltInProperties.PROPERTIES, Constraint.OTHER_OBJECT, OBJECT_PROTOTYPE);
+  public static final BuiltInConstructorSymbolicValue FUNCTION = new BuiltInConstructorSymbolicValue(FunctionBuiltInProperties.PROPERTIES, Constraint.FUNCTION, FUNCTION_PROTOTYPE);
+  public static final BuiltInConstructorSymbolicValue STRING = new BuiltInConstructorSymbolicValue(StringBuiltInProperties.PROPERTIES, Constraint.STRING_PRIMITIVE, STRING_PROTOTYPE);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java]: MINOR Split this 182 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java [hunk 3]
```diff
@@ -40,16 +41,27 @@
...
+  public static final BuiltInConstructorSymbolicValue NUMBER = new BuiltInConstructorSymbolicValue(NumberBuiltInProperties.PROPERTIES, Constraint.NUMBER_PRIMITIVE, NUMBER_PROTOTYPE);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java]: MINOR Split this 182 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java [hunk 4]
```diff
@@ -40,16 +41,27 @@
...
+  public static final BuiltInConstructorSymbolicValue BOOLEAN = new BuiltInConstructorSymbolicValue(BooleanBuiltInProperties.PROPERTIES, Constraint.BOOLEAN_PRIMITIVE, BOOLEAN_PROTOTYPE);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java]: MINOR Split this 186 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java [hunk 5]
```diff
@@ -40,16 +41,27 @@
...
-  NUMBER("Number", Type.NUMBER_OBJECT),
-  BOOLEAN("Boolean", Type.BOOLEAN_OBJECT),
-  STRING("String", Type.STRING_OBJECT),
-  FUNCTION("Function", Type.FUNCTION),
-  DATE("Date", Type.DATE),
-  REGEXP("RegExp", Type.REGEXP),
-  ARRAY("Array", Type.ARRAY),
-  OBJECT("Object", Type.OBJECT);
+  public static final BuiltInObjectSymbolicValue OBJECT_PROTOTYPE = new BuiltInObjectSymbolicValue(ObjectBuiltInProperties.PROTOTYPE_PROPERTIES, null, Constraint.OTHER_OBJECT);
+  public static final BuiltInObjectSymbolicValue FUNCTION_PROTOTYPE = new BuiltInObjectSymbolicValue(FunctionBuiltInProperties.PROTOTYPE_PROPERTIES, OBJECT_PROTOTYPE, Constraint.FUNCTION);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java]: MINOR Split this 188 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java [hunk 6]
```diff
@@ -80,9 +79,7 @@ public void afterBlockElement(ProgramState currentState, Tree element) {
   private static String getTypeOfObject(ProgramState programState) {
     Constraint constraintOnObject = programState.getConstraint(programState.peekStack());
     Type type = Type.find(constraintOnObject);
-    Preconditions.checkState(type != null && TYPE_NAMES.containsKey(type));
-
-    return TYPE_NAMES.get(type);
+    return TYPE_NAMES.containsKey(type) ? "a " + TYPE_NAMES.get(type) : "this object";
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java]: MAJOR Add parentheses to make the operator precedence explicit. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS864)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInProperty.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+import static org.sonar.javascript.se.Constraint.TRUTHY;
+
+abstract class BuiltInProperty {
+
+  public static final List<Constraint> ONE_NUMBER = ImmutableList.of(Constraint.ANY_NUMBER);
+  public static final List<Constraint> STRING_NUMBER = ImmutableList.of(Constraint.ANY_STRING, Constraint.ANY_NUMBER);
+  public static final List<Constraint> NUMBER_STRING = ImmutableList.of(Constraint.ANY_NUMBER, Constraint.ANY_STRING);
+  public static final List<Constraint> NUMBER_NUMBER = ImmutableList.of(Constraint.ANY_NUMBER, Constraint.ANY_NUMBER);
+  public static final List<Constraint> EMPTY = ImmutableList.of();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInProperty.java]: MINOR Make this member "protected". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/FunctionBuiltInProperties.java [hunk 8]
```diff
@@ -24,39 +24,27 @@
...
+    .put("apply", method(Constraint.ANY_VALUE, ImmutableList.of(Constraint.ANY_VALUE, Constraint.ANY_VALUE), true))
+    .put("bind", method(Constraint.FUNCTION, anyValues))
+    .put("call", method(Constraint.ANY_VALUE, anyValues, true))
+    // overrides Object
+    .put("toString", method(Constraint.STRING_PRIMITIVE, BuiltInProperty.EMPTY))
+    .put("length", property(Constraint.NUMBER_PRIMITIVE))
+    .put("name", property(Constraint.STRING_PRIMITIVE))
+    .build();
+
+  public static final Map<String, BuiltInProperty> PROPERTIES = ImmutableMap.of();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/FunctionBuiltInProperties.java]: MINOR Make this member "protected". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2386)

**Comment by SonarTech** [comment]
SonarQube analysis reported 12 issues
* MAJOR 1 major
* MINOR 11 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MINOR 3 minor

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java [hunk 9]
```diff
@@ -85,47 +54,32 @@
...
-    SymbolicValue valueForProperty = builtInProperties.getValueForProperty(propertyName);
-    if (valueForProperty == null) {
-      return getValueFromPrototype(propertyName);
+    if (prototype == null) {
+      throw new IllegalStateException("Cannot access a property on a " + this);
     }
-
-    return valueForProperty;
+    SymbolicValue propertyValue = prototype.getPropertyValue(propertyName);
+    if (!SpecialSymbolicValue.UNDEFINED.equals(propertyValue) || PRIMITIVE_TYPES.contains(this)) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: @pynicolas We still have pieces of logic for getting object property value spread over several classes: here, expression stack, BuiltInObjectSymbolicValue (there are even 2 methods: `getValueForProperty` and `getPropertyValue`). I think we should make an effort to concentrate it in one place

File: javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java [hunk 10]
```diff
@@ -72,58 +85,68 @@
...
-  public Type type() {
-    return type;
+  public static Optional<SymbolicValue> find(String name) {
+    return Optional.ofNullable(GLOBALS.get(name));
   }
 
-  @Override
-  public Optional<SymbolicValue> getValueForOwnProperty(String name) {
-    return type.getValueForOwnProperty(name);
+  public Constraint baseConstraint() {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/builtins/BuiltInObjectSymbolicValue.java]: @pynicolas I'm not sure about usage of this `baseConstraint`. Can we discuss that?

File: javascript-frontend/src/main/java/org/sonar/javascript/se/sv/InstanceOfSymbolicValue.java [hunk 11]
```diff
@@ -47,8 +48,10 @@ public InstanceOfSymbolicValue(SymbolicValue objectValue, SymbolicValue construc
   public Optional<ProgramState> constrainDependencies(ProgramState state, Constraint constraint) {
     Optional<ProgramState> newProgramState;
     Constraint constraintForObject = null;
-    if (constructorValue instanceof BuiltInObjectSymbolicValue) {
-      constraintForObject = ((BuiltInObjectSymbolicValue) constructorValue).type().constraint();
+    if (BuiltInObjectSymbolicValue.OBJECT.equals(constructorValue)) {
+      constraintForObject = Constraint.OBJECT;
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/sv/InstanceOfSymbolicValue.java]: @pynicolas Do we need this hack? what about having `base_constraint` equal `Constraint.OBJECT` and not `Constraint.OTHER_OBJECT` as in the current implementation?

**Pull request merged**